src/Orleans.Core/Streams/Internal/StreamConsumerExtension.cs

```diff
@@ -125,12 +125,16 @@ public Task<StreamHandshakeToken> DeliverImmutable(GuidId subscriptionId, Stream
             }
             else if(this.streamSubscriptionObserver != null)
             {
-                var subscriptionHandlerFactory = new StreamSubscriptionHandlerFactory(this.providerManager, streamId, streamId.ProviderName, subscriptionId);
-                await this.streamSubscriptionObserver.OnSubscribed(subscriptionHandlerFactory);
-                // check if an observer were attached after handling the new subscription, deliver on it if attached
-                if (allStreamObservers.TryGetValue(subscriptionId, out observer))
+                var streamProvider = this.providerRuntime.ServiceProvider.GetServiceByName<IStreamProvider>(streamId.ProviderName);
```

Comment:
should the results of this be cached somewhere?

Reply:
It also seems like 'GetObserver' logic could be split into a new method and shared between the two

Reply:
This should only be done on streams that do not already have an observer attached, and only for the first time that happens per stream.  While it's most likely that all streams on a grain will be on the same provider, multiple stream providers are supported, so we can't assume the same stream provider.  With this, I suspect a local cache would be of limited value.  The minimal possible perf gain would, imo, not justify any additional complexity.

Reply:
Fair enough, this code is still fairly foreign to me.

For example, I don't see how `allStreamObservers.TryGetValue(subscriptionId, out observer)` will ever evaluate to `false` above and then `true` below, but I suppose the idea is that the `IStreamSubscriptionObserver` has a chance to subscribe below.